crates/ra_ide/src/references/rename.rs

```diff
@@ -124,6 +127,116 @@ fn rename_mod(
     Some(SourceChange::from_edits("rename", source_file_edits, file_system_edits))
 }
 
+fn rename_to_self(db: &RootDatabase, position: FilePosition) -> Option<RangeInfo<SourceChange>> {
+    let sema = Semantics::new(db);
+    let source_file = sema.parse(position.file_id);
+    let syn = source_file.syntax();
+
+    let fn_def = find_node_at_offset::<ast::FnDef>(syn, position.offset)?;
+    let params = fn_def.param_list()?;
+    if params.self_param().is_some() {
+        return None; // method already has self param
+    }
+    let first_param = params.params().next()?;
+    let mutable = match first_param.ascribed_type() {
+        Some(ast::TypeRef::ReferenceType(rt)) => rt.mut_token().is_some(),
+        _ => return None, // not renaming other types
+    };
+
+    let RangeInfo { range, info: refs } = find_all_refs(db, position, None)?;
+
+    let param_range = first_param.syntax().text_range();
+    let (param_ref, usages): (Vec<Reference>, Vec<Reference>) = refs
+        .into_iter()
+        .partition(|reference| param_range.intersect(reference.file_range.range).is_some());
+
+    if param_ref.is_empty() {
+        return None;
+    }
+
+    let mut edits = usages
+        .into_iter()
+        .map(|reference| source_edit_from_reference(reference, "self"))
+        .collect::<Vec<_>>();
+
+    edits.push(SourceFileEdit {
+        file_id: position.file_id,
+        edit: TextEdit::replace(
+            param_range,
+            String::from(if mutable { "&mut self" } else { "&self" }),
+        ),
+    });
+
+    Some(RangeInfo::new(range, SourceChange::source_file_edits("rename", edits)))
+}
+
+fn source_edit_from_self(
+    syn: &SyntaxNode,
+    position: FilePosition,
```

Comment:
Though here it seems strange that depend on a specific position. It seems like this should be the same edit irrespective of which self we focus on, so perhaps this should take `ast::FnDef` instead?

Reply:
Edits are different because I'm also adding a type info when renaming self param. So that:
`f(&self)` gets replaced with `f(param: &Foo)`.
I can structure it differently, will make this helper function only handle self param, and do the simpler case (path) separately.